### PR TITLE
Add explicit limits to backend document selection parsing

### PR DESCRIPTION
**Pull request by vekterli**
@geirst please review

Adds the following (very generous) limits:
- Max AST depth of 1024
- Max input selection string size of 1 MiB

Have to track AST depth manually, as there is no exposed way of
doing this natively via Bison.

Also removed a regex that had the potential of catastrophic
backtracking in case of massive inputs. It wasn't removed during
the previous purge due to being used with capture groups, which are
not supported by our current vespalib regex wrapper.

### CODE REVIEW

File: document/src/vespa/document/select/parser.cpp [hunk 1]
```diff
@@ -8,7 +9,18 @@
 
 namespace document::select {
 
+namespace {
+
+void verify_expression_not_too_large(const std::string& expr) {
+    if (expr.size() > ParserLimits::MaxSelectionByteSize) {
+        throw ParsingFailedException("expression is too large to be parsed");
```
Comment (geirst) [inline document/src/vespa/document/select/parser.cpp:16]: Consider adding the size and limit to the exception message.

File: document/src/vespa/document/select/parser_limits.cpp [hunk 2]
```diff
@@ -0,0 +1,11 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+#include "parser_limits.h"
+#include "parsing_failed_exception.h"
+
+namespace document::select {
+
+void throw_max_depth_exceeded_exception() {
+    throw ParsingFailedException("expression is too deeply nested");
```
Comment (geirst) [inline document/src/vespa/document/select/parser_limits.cpp:8]: Consider adding the depth and max depth limit to the exception message.

**Review by geirst — APPROVED** [review]
👍  Consider adding some more details to the exception messages.

File: document/src/vespa/document/select/parser.cpp (hunk 1 above)
Comment (vekterli) [reply document/src/vespa/document/select/parser.cpp]: Done

File: document/src/vespa/document/select/parser_limits.cpp (hunk 2 above)
Comment (vekterli) [reply document/src/vespa/document/select/parser_limits.cpp]: Added limit to message. We cannot know the actual depth of the input AST as we terminate parsing immediately when hitting the limit.

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**